### PR TITLE
Fixed #32180 -- Added system check for file system caches absolute location.

### PR DESCRIPTION
**Pull request by cpina**
…s not absolute

### CODE REVIEW

File: django/core/checks/caches.py [hunk 1]
```diff
@@ -53,3 +54,24 @@ def check_cache_location_not_exposed(app_configs, **kwargs):
...
+    errors = []
+
+    if getattr(settings, 'CACHES', None):
+        for alias, config in settings.CACHES.items():
+            cache = caches[alias]
+
+            if not isinstance(cache, FileBasedCache):
+                continue
+
+            if not os.path.isabs(config['LOCATION']):
```
Comment (felixxm) [inline django/core/checks/caches.py:69]: You can use `Path.is_absolute()`.

File: django/core/checks/caches.py [hunk 2]
```diff
@@ -53,3 +54,24 @@ def check_cache_location_not_exposed(app_configs, **kwargs):
...
+                errors.append(Warning(
+                    f"Your '{alias}' LocalBasedCache LOCATION specifies a relative directory. "
+                    f"Use an absolute directory to make it independent of the working directory "
+                    f"of the Django process.",
+                    id='caches.W003'
+                ))
+
```
Comment (felixxm) [inline django/core/checks/caches.py:76]: Chop all blank lines in `check_file_based_cache_is_absolute()`.

File: django/core/checks/caches.py [hunk 3]
```diff
@@ -53,3 +54,24 @@ def check_cache_location_not_exposed(app_configs, **kwargs):
...
+            if not isinstance(cache, FileBasedCache):
+                continue
+
+            if not os.path.isabs(config['LOCATION']):
+                errors.append(Warning(
+                    f"Your '{alias}' LocalBasedCache LOCATION specifies a relative directory. "
+                    f"Use an absolute directory to make it independent of the working directory "
+                    f"of the Django process.",
+                    id='caches.W003'
```
Comment (felixxm) [inline django/core/checks/caches.py:74]:
```suggestion
                    id='caches.W003',
```

File: django/core/checks/caches.py [hunk 4]
```diff
@@ -53,3 +54,24 @@ def check_cache_location_not_exposed(app_configs, **kwargs):
...
+            cache = caches[alias]
+
+            if not isinstance(cache, FileBasedCache):
+                continue
+
+            if not os.path.isabs(config['LOCATION']):
+                errors.append(Warning(
+                    f"Your '{alias}' LocalBasedCache LOCATION specifies a relative directory. "
+                    f"Use an absolute directory to make it independent of the working directory "
+                    f"of the Django process.",
```
Comment (felixxm) [inline django/core/checks/caches.py:73]: Wrap at 79 chars.

File: django/core/checks/caches.py [hunk 5]
```diff
@@ -53,3 +54,24 @@ def check_cache_location_not_exposed(app_configs, **kwargs):
                 id='caches.W002',
             ))
     return errors
+
+@register(Tags.caches)
```
Comment (felixxm) [inline django/core/checks/caches.py:58]: I think it should be a `deploy` check.

File: django/core/checks/caches.py [hunk 6]
```diff
@@ -53,3 +54,24 @@ def check_cache_location_not_exposed(app_configs, **kwargs):
...
+    if getattr(settings, 'CACHES', None):
+        for alias, config in settings.CACHES.items():
+            cache = caches[alias]
+
+            if not isinstance(cache, FileBasedCache):
+                continue
+
+            if not os.path.isabs(config['LOCATION']):
+                errors.append(Warning(
+                    f"Your '{alias}' LocalBasedCache LOCATION specifies a relative directory. "
```
Comment (felixxm) [inline django/core/checks/caches.py:71]: Why `LocalBasedCache`? I think we can chop it.

File: tests/check_framework/test_caches.py [hunk 7]
```diff
@@ -89,3 +91,24 @@ def test_cache_path_not_conflict(self):
             settings = self.get_settings(setting, root / 'cache', root / 'other')
             with self.subTest(setting=setting), self.settings(**settings):
                 self.assertEqual(check_cache_location_not_exposed(None), [])
+
+class CheckCacheAbsolutePath(SimpleTestCase):
+    warning_message = (
+        "Your '%s' LocalBasedCache LOCATION specifies a relative directory. "
+        "Use an absolute directory to make it independent of the working directory "
+        "of the Django process."
+    )
```
Comment (felixxm) [inline tests/check_framework/test_caches.py:100]: `warning_message` is used in a single test, there is no need to add it to the class.

File: tests/check_framework/test_caches.py [hunk 8]
```diff
@@ -89,3 +91,24 @@ def test_cache_path_not_conflict(self):
...
+        self.assertEqual(check_file_based_cache_is_absolute(None), [])
+
+    @override_settings(CACHES={'default': {'BACKEND': 'django.core.cache.backends.filebased.FileBasedCache', 'LOCATION': '/tmp/django_cache'}})
+    def test_absolute_directory(self):
+        self.assertEqual(check_file_based_cache_is_absolute(None), [])
+
+    @override_settings(CACHES={'default': {'BACKEND': 'django.core.cache.backends.filebased.FileBasedCache', 'LOCATION': 'django_cache'}})
+    def test_relative_directory(self):
+        msg = self.warning_message % ('default', )
+        self.assertEqual(check_file_based_cache_is_absolute(None), [Warning(msg, id='caches.W003')])
```
Comment (felixxm) [inline tests/check_framework/test_caches.py:113]: There is no need to declare `warning_message` or `msg`:
```suggestion
        self.assertEqual(check_file_based_cache_is_absolute(None), [
            Warning(   
                "Your 'default' ...", 
                id='caches.W003',
            ),
        ])
```

File: django/core/checks/caches.py [hunk 9]
```diff
@@ -53,3 +54,24 @@ def check_cache_location_not_exposed(app_configs, **kwargs):
                 id='caches.W002',
             ))
     return errors
+
+@register(Tags.caches)
+def check_file_based_cache_is_absolute(app_configs, **kwargs):
+    errors = []
+
+    if getattr(settings, 'CACHES', None):
```
Comment (felixxm) [inline django/core/checks/caches.py:62]: IMO we don't need an extra check and `test_no_caches` test.

**Review by felixxm — COMMENTED** [review]
@cpina Thanks for this patch :+1: Please check my comments and fix `flake` and `isort`.

**Comment by felixxm** [comment]
@cpina Thanks for updates :+1: I squashed commits, fixed flake and isort, and pushed small edits.

**Review by cpina — COMMENTED** [review]
> @cpina Thanks for updates +1 I squashed commits, fixed flake and isort, and pushed small edits.

Thanks very much @felixxm - sorry that I missed `flake` and `isort` yesterday, I was reading the individual comments in detail and missed the general one. Next time!

File: django/core/checks/caches.py (hunk 5 above)
Comment (felixxm) [reply django/core/checks/caches.py]: I've changed it to a deploy check because it's not really an issue in a local environment.

Comment (felixxm) [reply django/core/checks/caches.py]: but OK let's have it.

**Pull request merged**